pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+
+    // V is the set of 2f+1 view change messages
+    for (const auto &sender_viewchange_envelope: viewchange_envelopes_from_senders)
+    {
+        *(newview.add_viewchange_messages()) = sender_viewchange_envelope.second;
+    }
+
+    // O
+    for (const auto& preprepare_message : pre_prepare_messages)
+    {
+        *(newview.add_pre_prepare_messages()) = preprepare_message.second;
+    }
+    return newview;
+}
+
+pbft_msg
+pbft::build_newview(uint64_t new_view, const std::map<uuid_t,bzn_envelope> viewchange_envelopes_from_senders) const
+{
+    //  Computing O (set of new preprepares for new-view message)
+    std::map<uint64_t, bzn_envelope> pre_prepares;
+
+    uint64_t max_checkpoint_sequence{0};
+    for (const auto& sender_viewchange_envelope : viewchange_envelopes_from_senders)
+    {
+        pbft_msg viewchange;
+        viewchange.ParseFromString(sender_viewchange_envelope.second.pbft());
+        max_checkpoint_sequence = std::max(max_checkpoint_sequence, viewchange.sequence());
+    }
+
+    //  - for each of the 2f+1 viewchange messages
+    for (const auto& sender_viewchange_envelope : viewchange_envelopes_from_senders)
+    {
+        //  -- for each operation included in that viewchange message
+        pbft_msg viewchange_message;
+        viewchange_message.ParseFromString(sender_viewchange_envelope.second.pbft());
+
+        for (int i{0}; i < viewchange_message.prepared_proofs_size(); ++i)
+        {
+            const auto& prepared_proof = viewchange_message.prepared_proofs(i);
+
+            //  --- if we have not already created a new preprepare for an operation with that sequence number
+
+            //  ---- Create a new preprepare for that operation using its original sequence number
+            //       and request hash, but using the new view number
+            pbft_msg pre_prepare;
+            pre_prepare.ParseFromString(prepared_proof.pre_prepare().pbft());
+            pre_prepare.set_view(new_view);
+
+            if (pre_prepare.sequence() <= max_checkpoint_sequence)
+            {
+                continue;
+            }
+            pre_prepares[pre_prepare.sequence()] = this->make_signed_envelope(pre_prepare.SerializeAsString());
+        }
+
+        //  - add each preprepare created this way to the new-view message
+    }
+
+    //  - for each sequence number between the base-sequence number (the last stable checkpoint)
+    //    and the highest sequence number for which a new preprepare was created
+
+    // Fill in missing messages with no ops
+    this->fill_in_missing_pre_prepares(max_checkpoint_sequence, new_view, pre_prepares);
+
+    //  - add each preprepare created this way to the new-view message
+    // std::set<std::string> view_change_messages
+    // viewchange_messages
+
+    return this->make_newview(new_view, viewchange_envelopes_from_senders, pre_prepares);
+}
+
+bzn_envelope
+pbft::make_signed_envelope(std::string serialized_pbft_message) const
```

Comment:
This should really be wrap_message. I understand that wrap_message doesn't currently sign the envelope, but it should (and will)

Reply:
I'll create a task to remove make_signed_envelope

Reply:
The reason *not* to change it now is that if we keep adding strictly uneccessary changes to this pull request , the pull request will never get done.